Start documenting database schema (#1446)
cowprotocol/services: database/README.md

With a live database information for all tables can be retrieved with the `\d` command and information for a specific table with `\d MyTable`.

The database contains the following tables:

### auction_participants

   Column    |  Type  | Nullable | Default
-------------|--------|----------|---------
 auction_id  | bigint | not null |
 participant | bytea  | not null |
Indexes:
- "auction_participants_pkey" PRIMARY KEY, btree (auction_id, participant)

This table is used for [CIP-20](https://snapshot.org/#/cow.eth/proposal/0x2d3f9bd1ea72dca84b03e97dda3efc1f4a42a772c54bd2037e8b62e7d09a491f). It stores which solvers (identified by ethereum address) participated in which auctions (identified by auction id). CIP-20 specifies that "solver teams which consistently provide solutions" get rewarded.

### auction_prices
### auction_transaction
### auctions (and auctions_id_seq counter))
### ethflow_orders
### ethflow_refunds
### fixed_bytes_test
### flyway_schema_history
### interactions
### invalidations
### onchain_order_invalidations
### onchain_placed_orders
### order_execution

Summary:  

> Contains metainformation for trades required to for reward computations that can not be recoverd from the block chain and are not stored in a persistent manner somewhere else.

Reviewer: ```suggestion
Contains metainformation for trades, required for reward computations that cannot be recovered from the blockchain and are not stored in a persistent manner somewhere else.
```

 Column      | Type    | Nullable | Default
-------------|---------|----------|--------
 order\_uid   | bytea   | not null |
 auction\_id  | bigint  | not null |
 reward      | double  | not null |
 surplus\_fee | numeric | nullable |
 solver\_fee  | numeric | nullable |

Indexes:  
- "order\_rewards\_pkey" PRIMARY KEY, btree (`order_uid`, `auction_id`)
  This table was originally called "order\_rewards" and renamed since then but the old index remained

Description:  
order\_uid: which order this trade execution is related to  
auction\_id: in which auction this trade was initiated  
reward: revert adjusted solver rewards, deprecated in favor of [CIP-20](https://snapshot.org/#/cow.eth/proposal/0x2d3f9bd1ea72dca84b03e97dda3efc1f4a42a772c54bd2037e8b62e7d09a491f)  
surplus\_fee: dynamic fee computed by the protocol that should get taken from the surplus of a trade, this value only applies and is set for fill-or-kill limit orders.  

> solver\_fee: value that is used for objective value computations. This either contains a fee equal to the execution cost of this trade computed by a solver (only applies to partially fillable limit orders) or the solver\_fee computed by the backend (which may include subsidies; applies to every other order) or it contains.  

Reviewer: Comment seems to not end properly (maybe typo?)

 liquiidty | These orders must be traded at their limit price and may not receive any surplus. Violating this is a slashable offence.
 limit     | Long lived order that may receive surplus. Users sign a static fee of 0 upfront and either the backend or the solvers compute a dynamic fee that gets taken from the surplus (while still respecting the user's limit price!).

Indexes:  
- "orders\_pkey" PRIMARY KEY, btree (`uid`)

Description:  
 uid: 56 bytes identifier composed of a 32 bytes `hash` over the order data signed by the user, 20 bytes containing the `owner` and 4 bytes containing `valid_to`.  
 owner: where the sell\_token will be taken from  
 creation\_timestamp: when the order was created  
 sell\_token: address of the token that will be sold  
 buy\_token: address of the token that will be bought  
 sell\_amount: amount in sell\_token that should at most be sold  
 buy\_amount: amount of buy\_token that should at least be bought  
 valid\_to: point in time when the order can no longer be settled  
 fee\_amount: amount in sell\_token the owner agreed upfront as a fee to be taken for the trade  
 kind: whether the order is a buy or a sell order  

>  partially\_fillable: determines if the order can be executed in multiple smaller trades or if it everything has to be executed at once  

Reviewer: ```suggestion
 partially\_fillable: determines if the order can be executed in multiple smaller trades or if everything has to be executed at once  
```

Here we document the current state of the database. The history of these changes lives in the `sql` folder which contains all migrations. This document shows the schema and the purpose of the tables.

Code that directly interfaces with the database through SQL queries lives in the `database`. This crate is often wrapped into higher level components by consumers.

With a live database information for all tables can be retrieved with the `\d` command and information for a specific table with `\d MyTable`.

The database contains the following tables:

### auction_participants

   Column    |  Type  | Nullable | Default
-------------|--------|----------|---------
 auction_id  | bigint | not null |
 participant | bytea  | not null |
Indexes:
- "auction_participants_pkey" PRIMARY KEY, btree (auction_id, participant)

This table is used for [CIP-20](https://snapshot.org/#/cow.eth/proposal/0x2d3f9bd1ea72dca84b03e97dda3efc1f4a42a772c54bd2037e8b62e7d09a491f). It stores which solvers (identified by ethereum address) participated in which auctions (identified by auction id).

> TODO: CIP-20 makes no mention of a need to know about all participants. Why do we store this?

Author: open question

Author: Is that documented more officially? It isn't part of CIP-20.

Reviewer 2: From the CIP:
```
The protocol commits to spend 20_000_000 COW tokens per year i.e. around 383307 COW per week on rewards. If at the end of an accounting period (currently: a week) the total amount of performance based rewards to be paid to solvers is smaller than the amount committed to by the protocol, the remaining sum is distributed to solver teams which consistently provide solutions. Those rewards will be distributed pro-rate to solvers based on the number of auctions they submitted a valid solution for over the accounting period.
```
